File: challs/broken_authentication.dir/broken_authentication.pl
#! /usr/bin/env perl

use strict;
use warnings;
use DBI;
use Digest::SHA  qw(sha1_hex);
use MIME::Base64;
use Mojolicious;
use Mojo::Log;
use Mojo::URL;
use Mojo::UserAgent;

my $driver = "SQLite";
my $dsn = "DBI:$driver:dbname=/tmp/broken_authentication.db";
my $dbh = DBI->connect($dsn, undef, undef, { sqlite_see_if_its_a_number => 1, AutoCommit => 1 }) or exit;

my $username = "";
my $userTokenMapping = "";
my $app = Mojolicious->new;
$app = $app->log(Mojo::Log->new(path => '/dev/null'));


sub generateUserToken {
    #
    #    Generates an API token for given username
    #
    my $username = shift;
    my $encoded = encode_base64($username, "");
    my $digest = sha1_hex($encoded);
    return $digest;
}

######################   API   ##############################

$app->hook(before_dispatch => sub {
    #
    #    Get token from HTTP header
    #
    my $c = shift;
    my $endpoint = $c->req->url->path;
   
    if ($endpoint eq '/get-token'){
        return;
    }
    
    my $token = $c->req->headers->header('X-API-TOKEN');
    my $sth = $dbh->prepare("SELECT username FROM users WHERE token=?");
    $sth->bind_param(1, $token);
    $sth->execute;
    $userTokenMapping = $sth->fetchrow_array;
    
    if (!$userTokenMapping) {
        return $c->render(
            status => 401,
            template => 'unauthorized',
            content => 'Invalid X-Api-Token'
         );
    }
});

$app->routes->get('/me' => sub {
    #
    #    Returns username based on api token value
    #
    my $c = shift;
    return $c->render(
        status => 200, 
        text => "Your token match the user '$userTokenMapping'"
    );
});

$app->routes->get('/get-token' => sub {
    #
    #    Create or update token for current user
    #
    my $c = shift;

    if ($username eq 'admin') {
        return $c->render(
            status => 403,
            template => 'forbidden',
            content => 'You are not allowed to request token for user admin'
        );
    }
    my $token = generateUserToken($username);
    my $sth = $dbh->prepare("INSERT OR REPLACE INTO users(username, token) VALUES(?, ?)");
    $sth->bind_param(1, $username);
    $sth->bind_param(2, $token);
    $sth->execute;
    
    return $c->render(
        status => 200,
        text => "Your API token is $token"
    );
});

$app->routes->get('/get-secret' => sub {
    #
    #    Return secret page
    #
    my $c = shift;
    if ($userTokenMapping ne 'admin') {
        return $c->render(
            status => 403,
            template => 'forbidden',
            content => "You are not allowed to request /get-secret"
        );
    }
    
    open my $fh, '<', 'secret' or die "error opening secret file";
    my $secret = do { local $/; <$fh> };
    return $c->render(
        status => 200, 
        text => "You are logged in. And congratz ! Here is the secret : $secret"
    );
});

####################     Main   #############################

sub main {

    $username = $ARGV[0];
    my $token = $ARGV[1];
    my $endpoint = $ARGV[2];

    if (!$username) {
        print "Missing username";
        exit 0;
    }
    
    if (!$endpoint) {
        $endpoint = '/get-token';
    }

    my $url = Mojo::URL->new($endpoint);
    my $ua = Mojo::UserAgent->new();
    $ua->server->app($app);
    my $response = $ua->get($url => {'X-API-TOKEN' => $token});
    print $response->res->content->asset->{'content'};
    $dbh->disconnect();
}

main();
1;

__DATA__

@@ forbidden.html.ep
% layout 'default';
<h1>Forbidden</h1>
<p><%= $content %></p>

@@ unauthorized.html.ep
% layout 'default';
<h1>Unauthorized</h1>
<p><%= $content %></p>

@@ not_found.html.ep
% layout 'default';
<h1>Not found</h1>

@@ layouts/default.html.ep
<!DOCTYPE html>
<html>
  <head><title>MyApp</title></head>
  <body><%= content %></body>
</html>

File: challs/broken_authentication.dir/broken_authentication.py
#!/usr/bin/python3

import sqlite3
import sys

from base64 import b64encode
from hashlib import sha1

from flask import Flask, g, request
from werkzeug.exceptions import (BadRequest, Forbidden,
                                 Unauthorized)


def get_params():
    """
        Check script params
    """
    if len(sys.argv) != 4:
        print("Missing parameters")
        sys.exit(0)

    username = sys.argv[1]
    token = sys.argv[2]
    endpoint = sys.argv[3]

    if not username:
        print("Missing username")
        sys.exit(0)

    return username, token, endpoint


def generate_user_token(username):
    """
        Generates an API token for given username
    """
    return sha1(b64encode(bytes(username, "utf-8"))).hexdigest()


def create_app():
    """
        Initialize Flask application
    """
    app = Flask(__name__)

    @app.before_request
    def get_user_from_token():
        """
            Get token from HTTP header
        """
        if request.path == '/get-token':
            return

        try:
            token = request.environ['HTTP_X_API_TOKEN']
        except (KeyError, IndexError, TypeError):
            raise BadRequest('Missing or invalid token')

        g.cursor.execute("SELECT username FROM users WHERE token=?", (token,))
        user = g.cursor.fetchone()
        if not user:
            raise Unauthorized("Invalid X-Api-Token")

        g.user_token_mapping = user[0]  # Store username in the request context

    @app.route('/me')
    def me():
        """
            Returns username based on api token value
        """
        return "Your token match the user '{}'".format(g.user_token_mapping)

    @app.route('/get-token')
    def get_token():
        """
            Create or update token for current user
        """
        if g.username == 'admin':
            raise Forbidden('You are not allowed to request token for user admin')

        token = generate_user_token(g.username)
        g.cursor.execute(
            "INSERT OR REPLACE INTO users(username, token) VALUES(?, ?)",
            (g.username, token)
        )
        return "Your API token is {}".format(token)

    @app.route('/get-secret')
    def get_secret():
        """
            Return secret page
        """
        if g.user_token_mapping != 'admin':  # Fetched from db in get_user_from_token() line #47
            raise Forbidden('You are not allowed to request /get-secret')

        with open('secret') as secret:
            return "You are logged in. And congratz ! Here is the secret : {}".format(
                secret.read()
            )

    return app


APP = create_app()
APP.config['DEBUG'] = True
APP.config['TESTING'] = True


if __name__ == '__main__':

    # Parse params
    username, token, endpoint = get_params()

    # Init app and db cursor
    try:
        tester = APP.test_client()
        ctx = APP.test_request_context()
        ctx.push()
        conn = sqlite3.connect('/tmp/broken_authentication.db', isolation_level=None)
        cursor = conn.cursor()
        g.cursor = cursor
        g.username = username
    except:
        print('Error while connecting to db.')
        sys.exit(1)

    if not endpoint:
        endpoint = '/get-token'

    # Make request
    response = tester.get(
        endpoint,
        headers={'X-API-TOKEN': token},
    )

    conn.commit()
    conn.close()
    print(response.get_data().decode('unicode_escape'))
